### ISSUE TITLE
Wine inside flatpak seems to lack something vulkan-related, perhaps DXVK?

### ISSUE THREAD

**Issue by ell1e** [flathub/io.github.sharkwouter.Minigalaxy#74]
Launching Doom 64's Windows build on Linux with minigalaxy inside the flatpak container gives the error "wine: Call from [some address] to unimplemented function dxgi.dll.CreateDXGIFactory, aborting". Googling around a bit seems to suggest wine implements this function under some circumstances, but it may require expansion libraries (but not native stuff with questionable licensing like winetricks) to be installed. Perhaps they could be added in, if somebody can figure out what it is? I'm guessing it might be either some vulkan-related things, or perhaps the DXVK components. I'm running into this with the 1.3.1 minigalaxy flatpak.

**Comment by sharkwouter** [comment]
Would be nice to fix this, but I personally am not sure how. Maybe someone else has more knowledge on this.

**Comment by ell1e** [comment]
Apparently, under ubuntu the process might be something like this:

```
sudo apt install -y dxvk
```
And then for each wineprefix:
```
export WINEPREFIX=<path to wineprefix that should be used>
dxvk-setup i
```
So I assume it's something that might need to be forwarded to whoever makes the flatpak runtime, if the `dxvk-setup` tool isn't already in the runtime.

**Comment by GB609** [comment]
The distribution-provided package is only a convenience to save space because it doesn't have to redownloaded and copied into every prefix. Instead the files are linked into the prefix when doing this.

But the easier way how DXVK can be installed would be to just use winetricks. It is not that large to begin with, so i personally think the convenience of usage is better than the amount of storage that can be saved by using some distribution-related package.

Either open winetricks for the game via properties in minigalaxy to get the winetricks ui and navigate the ui to select it, or execute the following command on the console (both require winetricks to be installed):

```sh
WINEPREFIX="/home/<username>/<gog-game-folder>/<doom-folder-name>/prefix" winetricks dxvk
```

It's important to have `prefix` as the last part of the WINEPREFIX because that is the actual folder name that minigalaxy uses as wineprefix when installing the game.

**Comment by ell1e** [comment]
My apologies if that's not a useful suggestion, but should minigalaxy perhaps be taught to run this? I imagine many people won't realize and will simply be left with a lot of the games not working.

**Comment by GB609** [comment]
There's no need for an apology.

The problem with wine and games is that the 'tricks' needed to run them can be vastly different. What works for one game can even completely break another one. That's one of the reasons why protonfixes and winetricks, protondb and winehq (or the customized lutris installers) exist the way they do.
Dxvk might be one of the less problematic tricks (but not always) to be added by default in terms of compatibility. However, there still are arguments against just including it by default.
- host driver support is not necessarily given.
- the dxvk-github itself warns that this could be considered cheating in multiplayer games as it is a manipulation of the direct3d libraries (from windows point of view)
- it would potentially mean adding another dependency to minigalaxy
- i personally had to find out the hard way that installing dxvk into a prefix is easier than cleanly removing it again. If we install it by default and it does break the game, we end up with broken default prefixes that must be 'repaired' manually by the user instead of being 'enhanced' with optional packages.
- version conflicts: some games might only work with a certain version of dxvk. There could also be an issue with combining some older driver versions newer dxvk version. Which one should minigalaxy pick? And how can it decide/find out?
- there's also the question of where to draw the line: if we integrate dxvk and some other game then requires something else, why shouldnt we include that other fix by default too? At the end of the day we would have to run something like 'install all winetricks' or maintain a database of fixes per game. The former would take a long time and wont work in all distributions because of the differences in 32bit/64bit prefixes. The latter is simply not feasible in terms of effort and is already done by lutris, steam, playOnLinux, umu etc.
Tldr: i think that minigalaxy by itself cant make sure that all windows games work all the time, theres just too much variation to maintain.

I also think that it's well documented that windows games don't always run out of the box in wine and that they might need tinkering. I'd rather say i would be suprised if the majority of linux gamers didnt know this (or learn of it very quickly). But thats just my gut feeling. Your guess is as good as mine though.

I think a better strategy here would be to keep that as is and continue the work to add support for umu to leverage (and ultimately maybe even contribute to) their common database (i've started this but got side-tracked/held up).

But thats only my personal opinion.

**Comment by ell1e** [comment]
I think a simple database of winetricks fixes may not be the worst idea if it's in some easily editable format like [JSONC](https://github.com/JSONC-Spec/JSONC-Spec?tab=readme-ov-file#common-jsonc-specification). Why not just let people make pull requests for fixes they found? If there's no entry with fixes then it'll just use e.g. a plain wine prefix.

Alternatively, perhaps the UI could at least allow checking a few of the most common fixes to apply. I imagine DXVK will be needed for a really large amount of modern bigger budget games.

**Comment by GB609** [comment]
> I think a simple database of winetricks fixes may not be the worst idea if it's in some easily editable format like [JSONC](https://github.com/JSONC-Spec/JSONC-Spec?tab=readme-ov-file#common-jsonc-specification). Why not just let people make pull requests for fixes they found? If there's no entry with fixes then it'll just use e.g. a plain wine prefix.
> 
> Alternatively, perhaps the UI could at least allow checking a few of the most common fixes to apply. I imagine DXVK will be needed for a really large amount of modern bigger budget games.

The 'winetricks' button in the game properties effectively is that UI you ask for.
Why duplicate effort that's been done somewhere else already?

Umu-Launcher is a collaboration of GloriousEggroll, Lutris, PlayOnLinux and Heroic Launcher (possibly more). The maintainers of all those tools realized that it's just duplicate effort if each has to maintain their own database for the same type of data and started to pool their efforts, see [UMU Database](https://github.com/Open-Wine-Components/umu-database).

I don't mean to disregard you, and i certainly understand where you are coming from. 
Some post-install event/script hook or UI for basic fixes might be a feasible middle-ground.
But i think that the effort needed to create and maintain yet another game fixes database is better spent elsewhere to improve minigalaxy and fix other bugs. Espcially because the wine version also is a moving target and otherwise varies between distributions in general. Some fixes we need today or in some wine version might not be needed tomorrow in another version and - on the contrary - instead start breaking prefixes. I wouldn't want to try and keep up with this.

Ultimately, it's just my opinion and not my call to make. @sharkwouter is making those decisions as the author of minigalaxy.
It's also a bit off-track for this issue, maybe we can find a better place to discuss about this. Maybe in your PR-to-be-made that introduces the 'common fixes ui' to minigalaxy?

**Comment by GB609** [comment]
https://github.com/doitsujin/dxvk/issues/4064
